Enable cross-compiling

froody opened the pull request:
Allow using and OS X build of swift to build libraries and binaries for another
platform, only linux-armv7 is supported in this patch. Doesn't currently support
building the tools (llvm, swiftc) or the tests for linux-armv7.
- Added --cross-compile-stdlib-targets to build-script
- Added --cross-compile-sysroot to build-script-impl
- Added --cross-compile-toolchain-bin to build-script-impl
- Set -resource-dir when cross-compiling
- Use BUILD_HOST_TARGET instead of HOST_TARGET, allow it to be empty when
  cross-compiling to just build cross target
- Derive PKG_CONFIG_PATH from --cross-compile-sysroot to allow find_package to
  work
- Skip cmark and llvm builds completely when --native-*-tools-path flags are set

See instructions in cmake/modules/Toolchain-linux-arm.cmake. To build
linux-armv7 on osx run:

`./utils/build-script -R --cross-compile-stdlib-targets linux-armv7 --
--cross-compile-sysroot=$MY_SYSROOT
--cross-compile-toolchain-bin=$MY_TOOLCHAIN_BIN`

gribozavr commented:
Why not use `debootstrap` instead of the ad-hoc make_sysroot script?

froody commented:
It was more work to figure out how to install/run debootstrap on OS X to do what I wanted, and provide easy-to-follow instructions, than to write this script. Would it be better if I just removed the script and hosted it myself, and suggest that users provide their own sysroot?

gribozavr commented:
debootstrap is the official ways to create sysroots for Debian and derivatives.  I'd rather we don't promote a hacky way.

gribozavr commented:
If someone needs to run debootstrap on OS X, they can easily use a VM, for example.

froody commented:
Ok, I removed my script, any thoughts on the rest of the patch?

Commit:
Enable cross-compiling

Allow using and OS X build of swift to build libraries and binaries for another
platform, only linux-armv7 is supported in this patch. Doesn't currently support
building the tools (llvm, swiftc) or the tests for linux-armv7.

* Added --cross-compile-stdlib-targets to build-script
* Added --cross-compile-sysroot to build-script-impl
* Added --cross-compile-toolchain-bin to build-script-impl
* Set -resource-dir when cross-compiling
* Use BUILD_HOST_TARGET instead of HOST_TARGET, allow it to be empty when
  cross-compiling to just build cross target
* Derive PKG_CONFIG_PATH from --cross-compile-sysroot to allow find_package to
  work
* Skip cmark and llvm builds completely when --native-*-tools-path flags are set

See instructions in cmake/modules/Toolchain-linux-arm.cmake. To build
linux-armv7 on osx run:

./utils/build-script -R --cross-compile-stdlib-targets linux-armv7 --
--cross-compile-sysroot=$MY_SYSROOT
--cross-compile-toolchain-bin=$MY_TOOLCHAIN_BIN

froody commented:
@gribozavr ping?

cellularmitosis commented:
I'm having a bit of trouble getting this pull request to work.

I've used your pull request as a patch against the 2/26 swift snapshot, and built up a set of scripts according to your instructions: https://github.com/cellularmitosis/swift-build-scripts

I'm running into the following error:

```
+ source versions.bash
++ export release=swift-DEVELOPMENT-SNAPSHOT-2016-02-25-a
++ release=swift-DEVELOPMENT-SNAPSHOT-2016-02-25-a
+ mkdir -p install products
+ cd stage
++ pwd
+ sysroot=/home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty
++ pwd
+ toolchain=/home/cell/swift.20160226.2/swift-build-scripts/stage/stage/gcc-arm-none-eabi-4_9-2015q3/bin
+ ./swift/utils/build-script -R --cross-compile-targets linux-armv7 -- --cross-compile-sysroot=/home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty --cross-compile-toolchain-bin=/home/cell/swift.20160226.2/swift-build-scripts/stage/stage/gcc-arm-none-eabi-4_9-2015q3/bin
Building the standard library for: swift-test-stdlib-linux-x86_64

find: `/home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty': No such file or directory
swift: using gold linker
+ cd /home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/swift-linux-armv7
+ /usr/bin/cmake -G Ninja -DCMAKE_C_COMPILER:PATH=/usr/bin/clang -DCMAKE_CXX_COMPILER:PATH=/usr/bin/clang++ -DCMAKE_TOOLCHAIN_FILE=/home/cell/swift.20160226.2/swift-build-scripts/stage/swift/cmake/modules/Toolchain-linux-arm.cmake -DLLVM_NATIVE_BUILD=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-x86_64/bin/../ -DCMAKE_SYSROOT:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty -DCMAKE_C_COMPILER_EXTERNAL_TOOLCHAIN:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/stage/gcc-arm-none-eabi-4_9-2015q3/bin -DSWIFT_ENABLE_GOLD_LINKER=TRUE -DLLVM_MAIN_SRC_DIR:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/llvm '-DCMAKE_C_FLAGS= -fno-stack-protector' '-DCMAKE_CXX_FLAGS= -fno-stack-protector' -DCMAKE_BUILD_TYPE:STRING=Release -DLLVM_ENABLE_ASSERTIONS:BOOL=TRUE -DSWIFT_ANALYZE_CODE_COVERAGE:STRING=FALSE -DSWIFT_STDLIB_BUILD_TYPE:STRING=Release -DSWIFT_STDLIB_ASSERTIONS:BOOL=TRUE -DSWIFT_STDLIB_ENABLE_RESILIENCE:BOOL=FALSE -DSWIFT_NATIVE_LLVM_TOOLS_PATH:STRING=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-x86_64/bin -DSWIFT_NATIVE_CLANG_TOOLS_PATH:STRING=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-x86_64/bin -DSWIFT_NATIVE_SWIFT_TOOLS_PATH:STRING=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/swift-linux-x86_64/bin -DSWIFT_BUILD_TOOLS:BOOL=FALSE -DSWIFT_BUILD_STDLIB:BOOL=TRUE -DSWIFT_SERIALIZE_STDLIB_UNITTEST:BOOL=FALSE -DSWIFT_BUILD_SDK_OVERLAY:BOOL=TRUE -DSWIFT_BUILD_STATIC_STDLIB:BOOL=FALSE -DSWIFT_BUILD_PERF_TESTSUITE:BOOL=FALSE -DSWIFT_BUILD_EXAMPLES:BOOL=TRUE -DSWIFT_INCLUDE_TESTS:BOOL=FALSE -DSWIFT_INSTALL_COMPONENTS:STRING= -DSWIFT_EMBED_BITCODE_SECTION:BOOL=FALSE -DSWIFT_ENABLE_LTO:BOOL=FALSE -DSWIFT_DARWIN_XCRUN_TOOLCHAIN:STRING=default -DSWIFT_AST_VERIFIER:BOOL=TRUE -DSWIFT_SIL_VERIFY_ALL:BOOL=FALSE -DSWIFT_RUNTIME_ENABLE_DTRACE:BOOL=FALSE -DSWIFT_RUNTIME_ENABLE_LEAK_CHECKER:BOOL=FALSE -DCMAKE_INSTALL_PREFIX:PATH=/usr -DLLVM_CONFIG:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-armv7/bin/llvm-config -DSWIFT_PATH_TO_CLANG_SOURCE:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/llvm/tools/clang -DSWIFT_PATH_TO_CLANG_BUILD:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-armv7 -DSWIFT_PATH_TO_LLVM_SOURCE:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/llvm -DSWIFT_PATH_TO_LLVM_BUILD:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-armv7 -DSWIFT_PATH_TO_CMARK_SOURCE:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/cmark -DSWIFT_PATH_TO_CMARK_BUILD:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/cmark-linux-armv7 -DSWIFT_CMARK_LIBRARY_DIR:PATH=/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/cmark-linux-armv7/src /home/cell/swift.20160226.2/swift-build-scripts/stage/swift
-- Using Linux-arm toolchain file
-- Using Linux-arm toolchain file
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang
-- Performing Test LLVM_NO_OLD_LIBSTDCXX
-- Performing Test LLVM_NO_OLD_LIBSTDCXX - Failed
CMake Error at /home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/llvm-linux-x86_64/share/llvm/cmake/HandleLLVMOptions.cmake:43 (message):
  Host Clang must be able to find libstdc++4.7 or newer!
Call Stack (most recent call first):
  cmake/modules/SwiftSharedCMakeConfig.cmake:231 (include)
  CMakeLists.txt:291 (swift_common_standalone_build_config)


-- Configuring incomplete, errors occurred!
See also "/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/swift-linux-armv7/CMakeFiles/CMakeOutput.log".
See also "/home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/swift-linux-armv7/CMakeFiles/CMakeError.log".
./swift/utils/build-script: command terminated with a non-zero exit status 1, aborting
```

Here's the contents of CMakeError.log:

```
Performing C++ SOURCE FILE Test LLVM_NO_OLD_LIBSTDCXX failed with the following output:
Change Dir: /home/cell/swift.20160226.2/swift-build-scripts/stage/build/Ninja-ReleaseAssert/swift-linux-armv7/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/ninja" "cmTryCompileExec758248503"
[1/2] Building CXX object CMakeFiles/cmTryCompileExec758248503.dir/src.cxx.o
FAILED: /usr/bin/clang++  --target=arm-linux-gnueabihf --sysroot=/home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty   -fno-stack-protector -B /home/cell/swift.20160226.2/swift-build-scripts/stage/stage/gcc-arm-none-eabi-4_9-2015q3/bin -B /home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8 -fuse-ld=gold -B /home/cell/swift.20160226.2/swift-build-scripts/stage/stage/gcc-arm-none-eabi-4_9-2015q3/bin -B /home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8 -fuse-ld=gold -DLLVM_NO_OLD_LIBSTDCXX -std=c++0x -isystem /home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty/usr/include/c++/4.8 -isystem /home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty/usr/include/arm-linux-gnueabihf/c++/4.8 -isystem /home/cell/swift.20160226.2/swift-build-scripts/stage/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8/include -o CMakeFiles/cmTryCompileExec758248503.dir/src.cxx.o -c src.cxx
src.cxx:2:10: fatal error: 'atomic' file not found
#include <atomic>
         ^
1 error generated.
ninja: build stopped: subcommand failed.

Source file was:

#include <atomic>
std::atomic<float> x(0.0f);
int main() { return (float)x; }
```

I have libstdc++ on the host system:

```
cell@hp(master)$ dpkg -l | grep libstdc++
ii  libstdc++-4.9-dev:amd64               4.9.2-10                             amd64        GNU Standard C++ Library v3 (development files)
ii  libstdc++-arm-none-eabi-newlib        4.8.3-9+4                            all          GNU Standard C++ Library v3 for ARM Cortex-A/R/M processors (newlib)
ii  libstdc++6:amd64                      4.9.2-10                             amd64        GNU Standard C++ Library v3
```

as well as in the sysroot:

```
cell@hp(master)$ find sysroots/sysroot.armhf.ubuntu.trusty/ | grep libstdc++
sysroots/sysroot.armhf.ubuntu.trusty/usr/share/doc/libstdc++-4.8-dev
sysroots/sysroot.armhf.ubuntu.trusty/usr/share/doc/libstdc++6
sysroots/sysroot.armhf.ubuntu.trusty/usr/share/doc/gcc-4.8-base/C++/README.libstdc++-baseline.armhf.gz
sysroots/sysroot.armhf.ubuntu.trusty/usr/share/doc/gcc-4.8-base/C++/changelog.libstdc++.gz
sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/arm-linux-gnueabihf/libstdc++.so.6.0.19
sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/arm-linux-gnueabihf/libstdc++.so.6
sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8/libstdc++.so
sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8/libstdc++.a
```

Am I missing something obvious?

froody commented:
Are you building this on a Linux machine? It looks like it from this log. I haven't accounted for that in this patch, but can. You need to add "linux-armv7" to the STDLIB_DEPLOYMENT_TARGETS for your host (see where STDLIB_DEPLOYMENT_TARGETS is first assigned in build-script-impl) and disable the host build somehow (an easy hacky way to see if it works is to comment out the "host" stdlib deployment target, so STDLIB_DEPLOYMENT_TARGETS=("linux-armv7"). I'll work on getting linux-x86_64 cross-compiling linux-armv7 into this patch too.

cellularmitosis commented:
That would be amazing if you could also work on linux-x86_64!  I'll also try your suggested workaround in the mean time.

gribozavr commented on utils/build-script:
```diff
@@ -302,6 +302,12 @@ details of the setups of other systems or automated environments.""")
+    targets_group.add_argument(
+        "--cross-compile-targets",
+        help="The cross-compliation targets. The Swift stdlib will be built for"
+             "these targets using the host-target LLVM and Clang",
```
This description does not sound correct to me.  We can't build the Swift standard library using LLVM and Clang, we need a Swift compiler.

gribozavr commented on utils/build-script:
```diff
@@ -552,6 +558,11 @@ also build for Apple watchos, but disallow tests that require an watchOS device"
 
+    parser.add_argument("--linux",
+        help="""
+also build for Linux""",
```
This description is unclear.  If we are on Linux, then we don't need a parameter.  If we are on a different OS, then `--cross-compile-targets` should be sufficient, right?

gribozavr commented on utils/build-script:
```diff
@@ -934,6 +950,22 @@ the number of parallel build jobs to use""",
     if args.extra_swift_args:
         build_script_impl_args += ["--extra-swift-args", ";".join(args.extra_swift_args)]
 
+    if args.cross_compile_targets:
```
If this option only cross-compiles the standard library, please rename it to make it clear.  As currently named, it makes it sound like it cross-compiles the whole compiler.

gribozavr commented on utils/build-script:
```diff
@@ -934,6 +950,22 @@ the number of parallel build jobs to use""",
+            "--native-llvm-tools-path=" + llvm_path,
+            "--native-clang-tools-path=" + llvm_path,
+            "--native-swift-tools-path=" + swift_path,
+            "--skip-build-cmark=1",
```
Does this require a two-step build-script invocation?  We should never require that.  The whole point of build-script is to orchestrate the whole build, taking care of the details.

froody replied:
Yeah, ok fixing that will take more refactoring

dabrahams commented:
FYI, I’m waiting to hear from @cellularmitosis that this has worked for him before proceeding to evaluate it.

gribozavr commented:
Invoking the build-script-impl twice from the top-level build-script is not a good solution either.  build-script-impl already has support for invoking cmake multiple times, it should be able to figure out everything.

cellularmitosis commented:
I'll fire off another build this evening.

Commit:
Move stdlib-deployment-targets calculation into build-script

* Demux uname info into a list of stdlib_deployment_targets in python
* Generate list of allowed stdlib_cross_targets based on host info

froody commented:
Updated to fix linux build.

@cellularmitosis: I successfully built on ubuntu 15.10/x86_64 with
`./utils/build-script -R --cross-compile-stdlib-targets linux-armv7 -- --reconfigure --cross-compile-sysroot=~/swifty/sysroot/armv7-sysroot/ --cross-compile-toolchain-bin=/usr/bin/arm-linux-gnueabi-`
can you try that with this latest patch? You may need to install gcc-arm-linux-gnueabi/

cellularmitosis commented:
Thanks!  Rebuilding.  I'll update this comment when the build finishes.

froody commented:
@cellularmitosis Can you change your script to set:
toolchain="`pwd`/stage/gcc-arm-none-eabi-4_9-2015q3/bin/arm-none-eabi-"
and delete the ./build/Ninja-ReleaseAssert/swift-linux-armv7/ directory and retry?

cellularmitosis commented:
Ah, now I understand.

I'll try again.

froody commented:
@cellularmitosis looks like your toolchain isn't set up correctly. On ubuntu I just did `apt-get install gcc-arm-linux-gnueabi` and then set `--cross-compile-toolchain-bin=/usr/bin/arm-linux-gnueabi-` since /usr/bin/arm-linux-gnueabi-ld.gold is installed by the above package. If you're not on ubuntu or your package manager doesn't install gold, you could download the [android NDK](http://developer.android.com/ndk/downloads/index.html) and find a copy of gold in there that works and copy it to `stage/gcc-arm-none-eabi-4_9-2015q3/bin/arm-none-eabi-ld.gold`

froody commented:
@cellularmitosis On ubuntu I just did `apt-get install gcc-arm-linux-gnueabi` and then set `--cross-compile-toolchain-bin=/usr/bin/arm-linux-gnueabi-` since /usr/bin/arm-linux-gnueabi-ld.gold is installed by the above package.

froody commented:
@cellularmitosis ping? This PR is blocked on you

cellularmitosis commented:
Hey @froody,

In the interest of speeding things up here, in an effort to more closely mirror the setup which has already worked for you, I ditched Debian in favor of Ubuntu 14.04, installed clang-3.6, gcc-arm-linux-gnueabi, and rebuilt.

I ran into the following error:

```
Run Build Command:/usr/bin/ninja cmTryCompileExec3944871647
[1/2] Building CXX object CMakeFiles/cmTryCompileExec3944871647.dir/src.cxx.o
[2/2] Linking CXX executable cmTryCompileExec3944871647
FAILED: : && /usr/bin/clang++-3.6   -fno-stack-protector -B /usr/bin/arm-linux-gnueabi- -B /home/cell/swift-build-scripts/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8 -fuse-ld=gold -B /usr/bin/arm-linux-gnueabi- -B /home/cell/swift-build-scripts/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8 -fuse-ld=gold -DLLVM_NO_OLD_LIBSTDCXX -std=c++0x   CMakeFiles/cmTryCompileExec3944871647.dir/src.cxx.o  -o cmTryCompileExec3944871647 -L/usr/lib/gcc/arm-linux-gnueabihf/4.8 -rdynamic && :
/usr/bin/arm-linux-gnueabi-ld.gold: error: /home/cell/swift-build-scripts/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8/crtbegin.o: incompatible target
/usr/bin/arm-linux-gnueabi-ld.gold: error: /home/cell/swift-build-scripts/sysroots/sysroot.armhf.ubuntu.trusty/usr/lib/gcc/arm-linux-gnueabihf/4.8/crtend.o: incompatible target
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.

Source file was:

#include <atomic>
std::atomic<float> x(0.0f);
int main() { return (float)x; }
```

froody commented:
@cellularmitosis I'm confused as to how you got this. I think what's missing is the "-target", but that should have been set by CMAKE_C_COMPILER_TARGET. I'm assuming you deleted the `swift-linux-armv7` dir? Can you paste the whole build log?

cellularmitosis commented:
@froody sure, after that failure I started a new build from scratch, to make sure the failure was reproducible.  I'll post the results when I get home from work today.

cellularmitosis commented:
@froody the from-scratch build failed:

Here's the entire build process output: https://gist.github.com/cellularmitosis/cc7fcb54a8e934c4df36

karwa commented:
I get the same error, on OSX trying to build for Debian using the sysroot/toolchain script in swift-build-scripts.

Also, in Toolchain-linux-arm.cmake:
`if (CMAKE_HOST_SYSTEM STREQUAL "Darwin")`
should be
`if (CMAKE_HOST_SYSTEM MATCHES "Darwin-*")`

because CMAKE_HOST_SYSTEM resolves to "Darwin-15.3" (on my MBP running El Cap 10.11.3)

modocache commented:
Cross-compilation is something I'm extremely interested in: #1157 and other armv7 pull requests took a long time to test because contributors had to compile for hours on underpowered machines to get test results. I'm very excited for this pull request--thanks, @froody!

I tried this out myself and managed to get everything working on OS X, following the steps in `make/modules/Toolchain-linux-arm.cmake`. I ran into two snags, though--here's how I fixed them on my end:

1. I followed the steps in the ran the [make_sysroot.py script](https://gist.github.com/froody/de6846f3455451f81992) linked to in the documentation. It failed initially because I was missing a `dpkg-deb` executable in my PATH. `brew install dpkg` got it working.
2. The documentation references `utils/build-script -R --cross-compile-targets`, but the correct option is now `--cross-compile-stdlib-targets`.

With the above I've successfully built the stdlib for linux-armv7 from an OS X machine:

```
$ file build/Ninja-ReleaseAssert/swift-linux-armv7/lib/swift/linux/libswiftCore.so
build/Ninja-ReleaseAssert/swift-linux-armv7/lib/swift/linux/libswiftCore.so: ELF 32-bit LSB shared object, ARM, version 1 (SYSV), dynamically linked (uses shared libs), not stripped
```

Exciting stuff! This would have made development on #1442 much simpler.

@dabrahams It'd be awesome if we could discuss whether this could be merged, I think it'd help a lot of people out (myself included).

gribozavr commented:
@modocache One thing that the PR is not doing correctly now is invoking build-script-impl twice.  build-script-impl already has support for cross-compilation, and this addition should fit into the existing scheme, so that we don't have two codepaths doing essentially the same, but slightly differently.

dabrahams commented:
@modocache I’ll take a look again after you’ve addressed that double-invocation-of-build-script issue

froody commented:
@karwa Can you give me steps to recreate your sysroot?

@gribozavr @dabrahams I've addressed the double-invocation-of-build-script issue and updated the commit message/PR description

karwa commented:
OK, I've got it to compile on OSX. I compiled a new toolchain, just grab the source from [here](https://launchpad.net/gcc-arm-embedded/+download) and edit the binutils build script to enable gold. I used the build output of that as my toolchain, but clang still couldn't resolve arm-none-eabi-ld.gold with the -fuse=gold option.

So I used the non-prefixed toolchain path (i.e. /arm-none-eabi/bin/ld.gold instead of /bin/arm-none-eabi-ld.gold) and that seemed to help it find everything. That means you need to change Toolchain-linux-arm.cmake and set AR path with "${CMAKE_C_COMPILER_EXTERNAL_TOOLCHAIN}/ar".

stefanstefan2001 commented:
What happend to this PR?

modocache commented:
@stefanstefan2001 I'm trying to split it up (see #2187) and get cross-compilation merged. It seems like a sticking point is the multiple invocations of `build-script-impl`, so I'll try to address that. I'm considering a different approach than using CMake toolchain files, but I'm still figuring out what's possible.

DougGregor commented:
This PR has been dormant for a long time. Is it still relevant?

dabrahams commented:
@DougGregor It's still relevant as long as build performance is cheaper to achieve on some platforms than on others.

dcartman commented:
Any update on when this might be merged?

slavapestov commented:
Can someone rebase this onto master and fix the conflicts, and we'll make sure to review it and get it merged ASAP?

karwa commented:
Some of the changes already got merged in. I've been maintaining a branch which contains the rest: https://github.com/karwa/swift

Basically the only things not yet merged are the CMake toolchain file and the pkg-config lookup in `build-script-impl`.

n8gray commented:
Does this PR still represent the state of cross-compiling in Swift?  If so, is there any documentation on how to use it?  I've been trying to build a cross-compiler to Raspberry Pi.  I've grabbed @karwa's fork from https://github.com/karwa/swift and rebased his patches to master, put together a toolchain and a sysroot, and built with this command:

```
./utils/build-script -r --stdlib-deployment-targets linux-armv7 \
  -- \
  --cross-compile-sysroots="/SAPDevelop/src/swiftX/sysroot.armhf.debian.jessie/" \
  --cross-compile-toolchains="/SAPDevelop/src/swiftX/toolchain.gcc-arm-none-eabi-5_2-2015q4/"
```

Eventually I got this error:
```
CMake Error at cmake/modules/SwiftUtils.cmake:14 (message):
  Error! Unknown SDKs: LINUX
Call Stack (most recent call first):
  CMakeLists.txt:763 (precondition)
```

I'd be grateful for pointers from anybody who's made this work.  Thanks!

CodaFi commented:
Since @karwa is maintaining a more up-to-date fork, and this pull request has fallen behind the more comprehensive efforts to enable cross-compilation, I'm going to close it.  Somebody should take a look at the state of the smaller pull requests that have been split off of this one and @karwa's branch and give us a more modern pull request.
